### DISCUSSION TITLE
How to use Windows.Services.Store with WinUI 3?

### DISCUSSION THREAD

**Discussion by Osaxely** [microsoft/WindowsAppSDK#3292] [Q&A]
Hello,
I am unable to use the Windows.Services.Store API with WinUI 3, but I cannot seem to be able to do anything.

I am unable to get the products add-ons (the app is associated with the Microsoft Store), the error 0x803F6107 still is thrown. I am also unable to use StoreContext.RequestRateAndReviewAppAsync(), the app crashes with the error "Access violation error".

Thanks in advance, any help is appreciated.

**Comment by DarranRowe** [comment]
Unfortunately, since you haven't given information on what you are actually doing, all I can do is point you at the documentation.

First, the [StoreContext](https://learn.microsoft.com/en-us/uwp/api/windows.services.store.storecontext?view=winrt-22621) documentation states that that it requires that you use [IInitializeWithWindow](https://learn.microsoft.com/en-us/windows/win32/api/shobjidl_core/nn-shobjidl_core-iinitializewithwindow). There should be a C# interop means of doing this if you are using C#.

Second, [Windows Runtime APIs not supported in desktop applications](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/desktop-to-uwp-supported-api#methods-that-use-the-request-naming-pattern) states that members with the Request naming pattern uses the Windows.UI.Popup class, and that means it requires a WinRT CoreWindow, which is just unavailable in desktop applications. Yes, WinUI 3 is still a desktop application. In this case, I don't think this means that you can't use it, but it means that you must use IInitializeWithWindow in order to tell the class to use a desktop window instead.

Third, [In-app purchases and trials](https://learn.microsoft.com/en-us/windows/uwp/monetize/in-app-purchases-and-trials#desktop) states that the StoreContext class must be configured to use a desktop HWND as the parent for modal dialogs. So this agrees with what I already stated above.

So to use StoreContext, you have to do the following steps.

1) Query the WinUI 3 window for the underlying HWND. This is done with the [IWindowNative](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/win32/microsoft.ui.xaml.window/nn-microsoft-ui-xaml-window-iwindownative) interface or the C# interop method. This must be obtained from the Xaml Window (Microsoft.UI.Xaml.Window) not any pages or user controls.

2) Create an instance of StoreContext.

3) Initialize it with the handle that you obtained from step 1.

To give an actual code examples, for C++
```
HWND window_handle = nullptr;
auto native_window = try_as<IWindowNative>();
winrt::check_hresult(native_window->get_WindowHandle(&window_handle));
winrt::Windows::Services::Store::StoreContext sc = nullptr;
//obtain store context instance
auto init_with_window = sc.as<IInitializeWithWindow>();
winrt::check_hresult(init_with_window->Initialize(m_window_handle));
```

For C#

```
Windows.Services.Store.StoreContext sc = null;
//obtain StoreContext instance
WinRT.Interop.InitializeWithWindow.Initialize(sc, WinRT.Interop.WindowNative.GetWindowHandle(this));
```

**Reply by Osaxely** [reply]
Hello, thank you for your response.
I have tried that, but unfortunately without success. This is what my class looks like:

```csharp
public void Initialize() {
  if (StoreContext == null) {
    // Get StoreContext for current user
    StoreContext context = StoreContext.GetDefault();

    // Get HWND
    IntPtr hwnd = WinRT.Interop.WindowNative.GetWindowHandle(App.Window); // App.Window is static, from App.xaml.cs, it's part of Microsoft.UI.Xaml

    // And set the HWND on the StoreContext
    WinRT.Interop.InitializeWithWindow.Initialize(context, hwnd);

    // Assign StoreContext
    StoreContext = context;
  }

   // ... Rest of the code
}
```

As soon as I invoke the popup to purchase something, nothing is opening, and the error ``0x803F6107`` is returned by the Microsoft Store services, which apparently is thrown when an app isn't associated with the Store, however it is associated with the store, and even published. As far as the StoreContext association goes, I am not sure if that is working. When I am using the method ``StoreContext.RequestRateAndReviewAppAsync()``, the app instantly crashes with ``AccessViolationException``. I do not know what's happening, but what I know is that this is happening in the method itself. I'll show you a screenshot showing the call stack and the error.

![image](https://user-images.githubusercontent.com/59126998/210169027-4105ae9e-af19-43fb-b21b-8981c51f1156.png)
![image](https://user-images.githubusercontent.com/59126998/210169035-45aa7b9a-ded3-4984-b75d-02763c570ecb.png)

Thanks in advance for your response.

**Reply by DarranRowe** [reply]
Unfortunately, there isn't much more to say here.

Since you are properly initialising the StoreContext object, then there are few possibilities left. Making the assumption that your code is correct and that you have the package properly registered with the store, which can't really be verified, all that is left is that this is either a documentation or a Windows issue.

But even if you got something wrong, RequestRateAndReviewAppAsync just outright crashing with a while trying to read a null pointer is bad and seriously indicates a bug somewhere. This is something that you would need to contact Windows developer support about.

**Reply by Enigma205** [reply]
What you did is wrong, you have to intialize the StoreContext reference not the class it self:

```
private StoreContext _storeContext;
private void Initialize()
{            
    // Retrieve the window handle (HWND) of the current WinUI 3 window.
    var hWnd = WinRT.Interop.WindowNative.GetWindowHandle(this);

    _storeContext = StoreContext.GetDefault();
    // Initialize the store context with the window handle (HWND).
    WinRT.Interop.InitializeWithWindow.Initialize(_storeContext, hWnd);
                
}
```

then on the UI thread call:
`
StoreRateAndReviewResult result = await  _storeContext.RequestRateAndReviewAppAsync();
`

the store rate and review will show so you can examine the result to do you action.

Ref: https://learn.microsoft.com/en-us/windows/apps/develop/ui-input/display-ui-objects
